internal/statemachine/common_states.go

```diff
@@ -133,66 +134,133 @@ func (stateMachine *StateMachine) generateDiskInfo() error {
+func (stateMachine *StateMachine) getRootfsMinSize() (quantity.Size, error) {
+	rootfsMinSize, err := helper.Du(stateMachine.tempDirs.rootfs)
+	if err != nil {
+		return quantity.Size(0), fmt.Errorf("Error getting rootfs size: %s", err.Error())
+	}
+
+	// Take into account ext4 filesystems metadata size
+	rootfsPadding := ext4Padding
+	rootfsMinSize = quantity.Size(math.Ceil(float64(rootfsMinSize) * ext4FudgeFactor))
+	rootfsMinSize += rootfsPadding
+
+	return stateMachine.alignToSectorSize(rootfsMinSize), nil
+}
+
+// getGadgetRootfsMinSize gets the minimum size of the rootfs requested in the
+// gadget YAML
+func (stateMachine *StateMachine) getGadgetRootfsMinSize(rootfsStructure *gadget.VolumeStructure) quantity.Size {
+	if rootfsStructure == nil {
+		return quantity.Size(0)
+	}
+
+	return stateMachine.alignToSectorSize(rootfsStructure.MinSize)
+}
+
+// getRootfsDesiredSize subtracts the size and offsets of the existing structures
+// from the requested image size to determine the maximum possible size of the rootfs
+func (stateMachine *StateMachine) getRootfsDesiredSize(rootfsVolume *gadget.Volume, rootfsVolumeName string) (quantity.Size, bool) {
+	if rootfsVolume == nil {
+		return quantity.Size(0), false
+	}
+
+	desiredSize, found := stateMachine.ImageSizes[rootfsVolumeName]
+	if !found {
+		// So far we do not know of a desired size for the rootfs
+		return quantity.Size(0), found
+	}
+
+	reservedSize := calculateNoRootfsSize(rootfsVolume)
+	// Create a temp partition table to get its size
+	// The value is already a multiple of the sector size, so no need to align again
+	tempPartitionTable := partition.NewPartitionTable(rootfsVolume, uint64(stateMachine.SectorSize), 0)
+	reservedSize += quantity.Size(tempPartitionTable.PartitionTableSize())
+
+	desiredSize = helper.SafeQuantitySubtraction(desiredSize, reservedSize)
+	desiredSize = stateMachine.alignToSectorSize(desiredSize)
+
+	return desiredSize, found
+}
+
+// calculateNoRootfsSize determines the needed space for existing structures
+// except for the rootfs
+func calculateNoRootfsSize(v *gadget.Volume) quantity.Size {
+	var size quantity.Size
+	for _, s := range v.Structure {
+		if helper.IsRootfsStructure(&s) { //nolint:gosec,G301
+			continue
+		}
+		if s.Offset != nil && quantity.Size(*s.Offset) > size {
+			size = quantity.Size(*s.Offset)
+		}
+		size += s.MinSize
+	}
+
+	return size
+}
+
+// findRootfsVolumeStructure finds the volume and the structure associated to the rootfs
+func (stateMachine *StateMachine) findRootfsVolumeStructure() (*gadget.Volume, string, *gadget.VolumeStructure) {
 	for volumeName, volume := range stateMachine.GadgetInfo.Volumes {
-		for _, structure := range volume.Structure {
-			if structure.Size == 0 {
-				return volume, volumeName
+		for i := range volume.Structure {
+			s := &volume.Structure[i]
+			if helper.IsRootfsStructure(s) { //nolint:gosec,G301
+				return volume, volumeName, s
```

Comment:
More tangents. But snapd's parsing of gadget.yaml sets volume.Name to volumeName here -- we don't need to pass volume names around with the volumes everywhere.

Reply:
Correct. Since we were looping on volumes and getting their name it was "free" to return it from `findRootfsVolumeStructure` but I supposed once we know the volume is not nil in `getRootfsDesiredSize` we can be reasonably sure it has a name.